### PR TITLE
smlfut: init at 1.1.0

### PR DESCRIPTION
**Pull request by athas**
## Description of changes

`smltfut` is a code generator that allows SML programs to call Futhark programs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Munksgaard** [comment]
I think the hash is off? When I tried building, it wanted `sha256-Ta0nCVD8N1k88sCdN4RhcugBgkQE7NdclCUtubgS6HM=` instead of `sha256-e0N8vQgR5Lte3eNiTu9d7O6m0i3zAPu67yMpy3HqAlo=`. 

After fixing that though, everything seems to work. Nice work!

**Comment by athas** [comment]
Should be fixed now.

**Comment by Munksgaard** [comment]
Lgtm :+1:

**Comment by sikmir** [comment]
Result of `nixpkgs-review pr 268106` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>smlfut</li>
  </ul>
</details>

File: pkgs/development/tools/smlfut/default.nix [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  };
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [ mlton ];
+
+  # Set as an environment variable in all the phase scripts.
+  MLCOMP = "mlton";
+
+  buildFlags = ["all"];
```
Comment (sikmir) [inline pkgs/development/tools/smlfut/default.nix:21]:
```suggestion
```

File: pkgs/development/tools/smlfut/default.nix [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  doCheck = true;
+
+  nativeCheckInputs = [ futhark ];
+
+  checkPhase = ''
+    runHook preCheck
+    make run_test
+    runHook postCheck
+  '';
```
Comment (sikmir) [inline pkgs/development/tools/smlfut/default.nix:32]:
```suggestion
  checkTarget = "run_test";
```

File: pkgs/development/tools/smlfut/default.nix [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  meta = with lib; {
+    description = "Generic package manager for Standard ML libraries and programs";
+    homepage = "https://github.com/diku-dk/smlfut";
+    license = licenses.gpl3Plus;
+    platforms = mlton.meta.platforms;
+    maintainers = with maintainers; [ athas ];
```
Comment (sikmir) [inline pkgs/development/tools/smlfut/default.nix:39]:
```suggestion
    maintainers = with maintainers; [ athas ];
    mainProgram = "smlfut";
```

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -17284,6 +17284,8 @@ with pkgs;
 
   smlpkg = callPackage ../tools/package-management/smlpkg { };
 
+  smlfut = callPackage ../development/tools/smlfut { };
```
Comment (sikmir) [inline pkgs/top-level/all-packages.nix:17287]: Consider moving to `pkgs/by-name/sm/smlpkg/package.nix` instead.

Comment (athas) [reply pkgs/top-level/all-packages.nix:17287]: Done.  I take it this is new best practice?  Fine by me; the old categories often seemed contrived anyway.

**Review by sikmir — APPROVED** [review]

File: pkgs/by-name/sm/smlfut/package.nix [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    rev = "v${version}";
+    hash = "sha256-Ta0nCVD8N1k88sCdN4RhcugBgkQE7NdclCUtubgS6HM=";
+  };
+
+  enableParallelBuilding = true;
+
+  nativeBuildInputs = [ mlton ];
+
+  # Set as an environment variable in all the phase scripts.
+  MLCOMP = "mlton";
```
Comment (emilytrau) [inline pkgs/by-name/sm/smlfut/package.nix:19]:
```suggestion
  env.MLCOMP = "mlton";
```

**Review by emilytrau — CHANGES_REQUESTED** [review]
Please squash your commits and use the standard format for the commit message -  `smlfut: init at 1.1.0`. Thanks!

File: pkgs/by-name/sm/smlfut/package.nix (hunk 5 above)
Comment (athas) [reply pkgs/by-name/sm/smlfut/package.nix]: Done.

**Comment by athas** [comment]
Squashed.

File: pkgs/by-name/sm/smlfut/package.nix (hunk 5 above)
Comment (emilytrau) [reply pkgs/by-name/sm/smlfut/package.nix]: Might have gotten lost in the squash :/

**Comment by athas** [comment]
All better now.

**Review by emilytrau — APPROVED** [review]

**Pull request merged**